JSON output [5]

Honny1 opened the pull request:
This PR creates the ability to use templates for `JSON` output. And creates `JSON-EVERYTHING` format which is `JSON` contains all information that `openscap-report` can offer.

Depends on PR: #103

evgenyz commented on openscap_report/report_generators/json.py:
```diff
@@ -0,0 +1,40 @@
+    def generate_report(self, debug_setting):
+        logging.warning("JSON Format is experimental output!")
+        template = self.env.get_template("base.json")
+        json_data = json.loads(template.render(report=self.report, debug_setting=debug_setting))
```
I'm not quite sure that Jinja is needed here. You can just filter out keys from the report dict and then export it to JSON. Even if in the future there would be some more sophisticated logic (like keeping report.foo.bar and report.foo.baz but not report.foo.others) it'd be easier to just write a more complicated filter function rather than to siphon the whole report through Jinja engine.

Honny1 replied:
You're right.

evgenyz commented on openscap_report/scap_results_parser/parsers/info_of_test_parser.py:
```diff
@@ -101,17 +102,27 @@ def get_object(self, id_object):
+        self._transform_tuples_to_dict(out)
+        if out:
+            return out
+        return None
```
First of all you can just `return out or None`.

But I don't quite understand why you're trying to use None instead of an empty list if there are no elements. What benefit does it give? I can see a drawback, though. In situations when one would iterate over a list there always should be a check if the list is not None.

Honny1 replied:
Usage `None` instead empty array. I want to state that there is no data in the JSON for the given item. A dictionary filter could solve this.

evgenyz replied:
Sure, but why do you think that
```
foo {
   bars_list: None
}
```
is better than
```
foo {
   bars_list: []
}
```
?

Honny1 replied:
In terms of python code, it is better to use an empty array. As for the JSON content, I think it's better to use null because it's more obvious that there's no data or this data is optional. I think it is about a discussion about JSON content. Which is the best value to use for cases when there is no data?

evgenyz replied:
You're right, let's discuss it in the JSON structure issue. But here is some homework for you: https://www.infoq.com/presentations/Null-References-The-Billion-Dollar-Mistake-Tony-Hoare/ :)

evgenyz commented on openscap_report/scap_results_parser/parsers/rule_parser.py:
```diff
@@ -1,51 +1,53 @@
+            references.append(Reference(referenc.get("href"), referenc.text))
+        if references:
+            return references
+        return None
```
None instead of [], why?

evgenyz commented on openscap_report/scap_results_parser/parsers/rule_parser.py:
```diff
@@ -1,51 +1,53 @@
+            identifiers.append(Identifier(identifier.get("system"), identifier.text))
+        if identifiers:
+            return identifiers
+        return None
```
None instead of [], why?

evgenyz commented on openscap_report/scap_results_parser/parsers/rule_parser.py:
```diff
@@ -1,51 +1,53 @@
-        return warnings
+        if warnings:
+            return warnings
+        return None
```
None instead of [], why?

evgenyz commented on openscap_report/scap_results_parser/parsers/rule_parser.py:
```diff
@@ -1,51 +1,53 @@
-        return output
+        if output:
+            return output
+        return None
```
None instead of [], why?

evgenyz commented on openscap_report/scap_results_parser/parsers/rule_parser.py:
```diff
@@ -80,17 +84,98 @@ def process_rule(self, rule):
+    def get_rules(self):
+        rules = {}
+        for rule in self.root.findall(".//xccdf:Rule", NAMESPACES):
+            rule = self.process_rule(rule)
```
This kind of shadowing could lead to a lot of troubles in a dynamically typed language. If you can have a Rule class entity and a rule xml sub-document it's better to introduce a naming scheme which won't allow you to mix them up. For example for all xml entities it could be `_el`s (`rule_el`).

Honny1 commented:
@evgenyz  I changed the code according to your review. And I dropped the json-output-2 branch because there was code there that replace empty arrays with none.

Honny1 commented:
/packit retest-failed

Honny1 commented:
/packit copr-build
